### PR TITLE
feat(runtime): SHA3-256 FIPS 202 and ECRecover Uncompressed Public key precompile

### LINKED ISSUE #4285: Feature Request: SHA3-256 FIPS 202 and ECRecover Uncompressed Public key precompile to Math API
1.  1.  1.  Description
             As part of enabling interoperability between NEAR Protocol and ICON Blockchain we would need to implement SHA3-256 FIPS 202 hash precompile and ECRecover Uncompressed public key precompile to Runtime Logics.

1.  1.  1.  Rationale

*   In order to validate the merkle tree data from ICON, the SHA3-256 FIPS 202 hashing function is required because all of ICON blockchain's hash and transaction hash are created using the SHA3-256 FIPS 202 hashing function
*   In order to validate signatures of validator of ICON, It is needed to recover the uncompressed public key from secp256k1 elliptic curve signature.

1.  1.  1.  Example
             Test cases done to check the input and output of the functions for the given parameters below

sha3_256
 params
 value : ‘0x0448250ebe88d77e0a12bcf530fe6a2cf1ac176945638d309b840d631940c93b78c2bd6d16f227a8877e3f1604cd75b9c5a8ab0cac95174a8a0a0f8ea9e4c10bca’
 returns : ‘0xc7647f7e251bf1bd70863c8693e93a4e77dd0c9a689073e987d51254317dc704’

ecrecover_public_key
 params
 hash : ‘0xc5d6c454e4d7a8e8a654f5ef96e8efe41d21a65b171b298925414aa3dc061e37’
 v : ‘0x00’
 r : ‘0x4011de30c04302a2352400df3d1459d6d8799580dceb259f45db1d99243a8d0c’
 s : ‘0x64f548b7776cb93e37579b830fc3efce41e12e0958cda9f8c5fcad682c610795’
 returns : ‘0x0448250ebe88d77e0a12bcf530fe6a2cf1ac176945638d309b840d631940c93b78c2bd6d16f227a8877e3f1604cd75b9c5a8ab0cac95174a8a0a0f8ea9e4c10bca’

1.  1.  1.  Changes

*   Use <span class="error">[tiny_keccak]</span>([https://docs.rs/tiny-keccak/2.0.2/tiny_keccak/](https://docs.rs/tiny-keccak/2.0.2/tiny_keccak/)) to implement SHA3 256 FIPS 202
*   Use <span class="error">[libsecp256k1]</span>([https://docs.rs/libsecp256k1/0.3.5/secp256k1/](https://docs.rs/libsecp256k1/0.3.5/secp256k1/)) to recover uncompressed public key
*   Implement precompile as Host Function API to VM Logics
*   Add Gas cost for the implementations
*   Implement necessary tests and test contracts
*   Feature Guard the newly introduced implementations

1.  1.  1.  References
             [https://csrc.nist.gov/publications/detail/fips/202/final](https://csrc.nist.gov/publications/detail/fips/202/final)
             [https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm#Public_key_recovery](https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm#Public_key_recovery)

### PR DESCRIPTION
**Pull request by MuhammedIrfan**
### Description
As part of enabling interoperability between NEAR Protocol and ICON Blockchain we would need to implement SHA3-256 FIPS 202 hash precompile and ECRecover Uncompressed public key precompile to Runtime Logics.

### Rationale
* In order to validate the merkle tree data from ICON, the SHA3-256 FIPS 202 hashing function is required because all of ICON blockchain's hash and transaction hash are created using the SHA3-256 FIPS 202 hashing function
* In order to validate signatures of validator of ICON, It is needed to recover the uncompressed public key from secp256k1 elliptic curve signature.

### Example
Test cases done to check the input and output of the functions for the given parameters below

sha3_256
params
value : ‘0448250ebe88d77e0a12bcf530fe6a2cf1ac176945638d309b840d631940c93b78c2bd6d16f227a8877e3f1604cd75b9c5a8ab0cac95174a8a0a0f8ea9e4c10bca’
returns : ‘c7647f7e251bf1bd70863c8693e93a4e77dd0c9a689073e987d51254317dc704’

ecrecover_public_key
params
hash : ‘c5d6c454e4d7a8e8a654f5ef96e8efe41d21a65b171b298925414aa3dc061e37’
v : ‘00’
r : ‘4011de30c04302a2352400df3d1459d6d8799580dceb259f45db1d99243a8d0c’
s : ‘64f548b7776cb93e37579b830fc3efce41e12e0958cda9f8c5fcad682c610795’
returns : ‘0448250ebe88d77e0a12bcf530fe6a2cf1ac176945638d309b840d631940c93b78c2bd6d16f227a8877e3f1604cd75b9c5a8ab0cac95174a8a0a0f8ea9e4c10bca’

### Libraries Used
* Use [tiny_keccak](https://docs.rs/tiny-keccak/2.0.2/tiny_keccak/) to implement SHA3 256 FIPS 202
* Use [libsecp256k1](https://docs.rs/libsecp256k1/0.3.5/secp256k1/) to recover uncompressed public key

### Issues
 * Closes #4285
 
### References
https://csrc.nist.gov/publications/detail/fips/202/final
https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm#Public_key_recovery

### CODE REVIEW

File: runtime/near-vm-logic/src/logic.rs [hunk 1]
```diff
@@ -951,6 +951,79 @@ impl<'a> VMLogic<'a> {
...
+        let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
+        self.gas_counter.pay_per_byte(sha3_256_byte, value.len() as u64)?;
+
+        let mut value_hash = vec![0; 32];
+        let mut sha3 = tiny_keccak::Sha3::v256();
+
+        sha3.update(&value);
+        sha3.finalize(&mut value_hash);
+
+        self.internal_write_register(register_id, value_hash.to_vec())
```
Comment (matklad) [inline runtime/near-vm-logic/src/logic.rs:978]:
```suggestion
        self.internal_write_register(register_id, value_hash)
```

File: runtime/near-vm-logic/src/logic.rs [hunk 2]
```diff
@@ -951,6 +951,79 @@ impl<'a> VMLogic<'a> {
...
+        let hash = self.memory_get_vec(hash_ptr, 32)?;
+        let v = (v & 0xFF) as u8;
+        let r = self.memory_get_vec(r_ptr, 32)?;
+        let s = self.memory_get_vec(s_ptr, 32)?;
+
+        let hash = secp256k1::Message::parse_slice(hash.as_slice()).unwrap();
+
+        let mut signature = [0u8; 64];
+        signature[0..32].copy_from_slice(r.as_slice());
+        signature[32..64].copy_from_slice(s.as_slice());
```
Comment (matklad) [inline runtime/near-vm-logic/src/logic.rs:1011]: Probably makes sense to add a comment why .unwrap() and `copy_from_slice` can't panic.

File: runtime/near-vm-logic/tests/test_miscs.rs [hunk 3]
```diff
@@ -530,6 +530,72 @@ fn test_keccak512() {
...
+        ExtCosts::write_register_base: 1,
+        ExtCosts::write_register_byte: 32,
+        ExtCosts::sha3_256_base: 1,
+        ExtCosts::sha3_256_byte: len,
+    });
+}
+
+#[test]
+#[cfg(feature = "protocol_feature_btp")]
+fn test_ecrecover_public_key() {
```
Comment (matklad) [inline runtime/near-vm-logic/tests/test_miscs.rs:567]: Let's add test for the two error cases as well.

File: runtime/near-vm-logic/src/logic.rs [hunk 4]
```diff
@@ -951,6 +951,79 @@ impl<'a> VMLogic<'a> {
...
+        let signature = secp256k1::Signature::parse(&signature);
+
+        let recovery_id = match secp256k1::RecoveryId::parse(v) {
+            Err(_) => return Err(HostError::InvalidECDSASignature.into()),
```
Comment (matklad) [inline runtime/near-vm-logic/src/logic.rs:1015]: Hm, I wonder what's the right behavior here. My understanding is that errors on this layer lead to contract termination with an error. So, if someone sends an invalid signature for the contract, the contract won't be able to recover from this. Is this the behavior we want? (I don't know myself, but superficially it seems that this is the first function which has this kind of failure mode).

**Comment by bowenwang1996** [comment]
Protocol changes, especially those that touch host functions, need to go through the quality control process https://gov.near.org/t/quality-control-for-protocol-api-changes/1941

File: runtime/near-vm-logic/src/logic.rs (hunk 4 above)
Comment (MuhammedIrfan) [reply runtime/near-vm-logic/src/logic.rs:1015]: @matklad You are right. As per the current behavior, the contract consuming this API will fail which ideally shouldn't be the case and error handling should be handled explicitly in the contract.

**Comment by MuhammedIrfan** [comment]
> Protocol changes, especially those that touch host functions, need to go through the quality control process https://gov.near.org/t/quality-control-for-protocol-api-changes/1941

Sure. We will follow the process.

**Comment by MuhammedIrfan** [comment]
@bowenwang1996 I have added a proposal https://gov.near.org/t/add-sha3-256-fips-202-to-math-api/2469 . Kindly check it. Thank you

**Pull request closed without merging**